src/crypto/tls/ech.go

```diff
@@ -161,49 +180,53 @@ func (c *Conn) echOfferOrBypass(helloBase *clientHelloMsg) (hello, helloInner *c
 	hello = new(clientHelloMsg)
 	*hello = *helloBase
 
+	// Produce covertext for the ECH extension.
+	//
+	// Using a hard-coded KEM public key, generate a fresh encapsulated key
+	// "enc". Generate a random "config_id" and "payload" of appropriate length.
 	hpkeSuite, err := hpkeAssembleCipherSuite(dummyKemId, dummyKdfId, dummyAeadId)
 	if err != nil {
 		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
 	}
-
 	pk, err := hpkeSuite.KEM.Deserialize(echDummyX25519PublicKey)
 	if err != nil {
 		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
 	}
 
+	// Compute the dummy context handle ("cipher_suite", "config_id", and
+	// "enc").
 	var dummyEch echClient
+	dummyEch.handle.suite.kdfId = dummyKdfId
+	dummyEch.handle.suite.aeadId = dummyAeadId
 	if !c.hrrTriggered {
-		dummyEch.handle.suite.kdfId = dummyKdfId
-		dummyEch.handle.suite.aeadId = dummyAeadId
 		dummyEch.handle.configId = make([]byte, dummyKdfOutputLen)
 		if _, err = io.ReadFull(config.rand(), dummyEch.handle.configId); err != nil {
 			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
 		}
-		c.ech.hrrCipherSuite = dummyEch.handle.suite
 		c.ech.hrrConfigId = dummyEch.handle.configId
 	} else {
 		// After HRR, the server checks these fields match.
-		dummyEch.handle.suite = c.ech.hrrCipherSuite
 		dummyEch.handle.configId = c.ech.hrrConfigId
 	}
-
 	dummyEch.handle.enc, _, err = hpke.SetupBaseS(hpkeSuite, config.rand(), pk, nil)
 	if err != nil {
 		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
 	}
 
-	// TODO(cjpatton): Ensure the length of the dummy ciphertext doesn't stick
-	// out.
-	dummyEch.payload = make([]byte, len(hello.marshal())+dummyAeadOverheadLen)
+	// Compute the dummy "payload".
+	dummyEncodedHelloInnerLen := 100 // TODO(cjpatton): Compute this correctly.
```

Comment:
You would do something like that, yeah. I didn't do it here to keep things a bit simpler. I thought it could wait until we do a more thorough "don't stick out" review in a future draft (probably -09). I'm happy to implement it now, if you prefer.

Reply:
The TODO is fine for now. :)